File: Data Structures & Algorithms/binary-tree-zigzag-level-order-traversal/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        queue = [root]
        result = []
        while queue:
            print("nw")
            res = []
            curr_res = []
            for i in queue:
                curr = i
                # print(curr.val)
                curr_res.append(curr.val)
                if i.left: res.append(i.left)
                if i.right: res.append(i.right)
            queue = []
            queue.extend(res)
            result.append(curr_res)

        for i in range(len(result)):
            if i % 2 == 1:
                result[i] = reversed(result[i])

        return result

File: Data Structures & Algorithms/binary-tree-diameter/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int width = 0;
    int solve(TreeNode* root, int depth){
        if(root == NULL) return 0;        
        int l = solve(root->left, depth + 1);
        std::cout << root->val << " " << depth <<std::endl;
        int r = solve(root->right, depth + 1);
        width = max(width, l + r);
        std::cout << "l:" << l << " " << "r:" << r << std::endl;
        return max(l,r) + 1;
    }
    int diameterOfBinaryTree(TreeNode* root) {
        int res = solve(root, 0);
        return width;
    }
};
